src/Microsoft.AspNetCore.Server.Kestrel/Http/Frame.cs

```diff
@@ -170,21 +170,23 @@ public bool HasResponseStarted
             get { return _requestProcessingStatus == RequestProcessingStatus.ResponseStarted; }
         }
 
-        protected FrameRequestHeaders FrameRequestHeaders => _frameHeaders.RequestHeaders;
+        protected FrameRequestHeaders FrameRequestHeaders { get; private set; }
+
+        protected FrameResponseHeaders FrameResponseHeaders { get; private set; }
 
         public void InitializeHeaders()
         {
-            if (_frameHeaders == null)
+            if (FrameRequestHeaders == null)
             {
-                _frameHeaders = new Headers(ServerOptions);
-                RequestHeaders = _frameHeaders.RequestHeaders;
-                ResponseHeaders = _frameHeaders.ResponseHeaders;
+                RequestHeaders = FrameRequestHeaders = new FrameRequestHeaders();
```

Comment:
@halter73 largely because everything else is. in theory you could wrap the headers collection and apply filters/transforms. No, I would not expect Kestrel to send headers from a new dictionary.